### PR TITLE
Introduce relaxed accuracy thresholds for CL target in some dnn tests.

### PR DESCRIPTION
**Pull request by aarongreig**
Partially addresses #9821

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/test/test_halide_layers.cpp [hunk 1]
```diff
@@ -258,7 +258,18 @@ TEST_P(LRN, Accuracy)
...
+    // The OpenCL kernels use the native_ math functions which have
+    // implementation defined accuracy, so we use relaxed thresholds. See
+    // https://github.com/opencv/opencv/issues/9821 for more details.
+    if (targetId == DNN_TARGET_OPENCL)
+    {
+        test(lp, input, backendId, targetId, false, 0.01f, 0.01f);
+    }
+    else
+    {
+        test(lp, input, backendId, targetId, false);
```
Comment (alalek) [inline modules/dnn/test/test_halide_layers.cpp:271]: Please use scheme for check tolerance adjusting similar to `test_torch_importer.cpp` file changes.

File: modules/dnn/test/test_layers.cpp [hunk 2]
```diff
@@ -169,8 +169,19 @@ TEST_P(Test_Caffe_layers, Softmax)
...
+    // https://github.com/opencv/opencv/issues/9821 for more details.
+    if (target == DNN_TARGET_OPENCL)
+    {
+        testLayerUsingCaffeModels("layer_lrn_spatial", false, true, 0.01f, 0.01f);
+        testLayerUsingCaffeModels("layer_lrn_channels", false, true, 0.01f, 0.01f);
+    }
+    else
+    {
+        testLayerUsingCaffeModels("layer_lrn_spatial", false, true);
+        testLayerUsingCaffeModels("layer_lrn_channels", false, true);
```
Comment (alalek) [inline modules/dnn/test/test_layers.cpp:183]: Please use scheme for check tolerance adjusting similar to `test_torch_importer.cpp` file changes.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

It makes sense to put this patch to [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

Please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

File: modules/dnn/test/test_torch_importer.cpp [hunk 3]
```diff
@@ -218,9 +218,26 @@ TEST_P(Test_Torch_layers, net_conv_gemm_lrn)
...
-    runTorchNet("net_conv_gemm_lrn", "", false, true, true,
-                target == DNN_TARGET_OPENCL_FP16 ? 0.046 : 0.0,
-                target == DNN_TARGET_OPENCL_FP16 ? 0.023 : 0.0);
+    double l1 = 0.0, lInf = 0.0;
+    if (target == DNN_TARGET_OPENCL_FP16)
+    {
+        l1 = 0.046;
+        lInf = 0.023;
+    }
+    else if (target == DNN_TARGET_CUDA_FP16)
```
Comment (aarongreig) [inline modules/dnn/test/test_torch_importer.cpp:227]: wasn't sure what to do about backporting the cuda check since it wasn't in the file in 3.4

Comment (alalek) [reply modules/dnn/test/test_torch_importer.cpp:227]: Right, please remove this code block from the patch:
```
    else if (target == DNN_TARGET_CUDA_FP16)
    {
        l1 = 0.0042;
        lInf = 0.021;
    }
```

File: modules/dnn/test/test_halide_layers.cpp [hunk 4]
```diff
@@ -251,7 +251,17 @@ TEST_P(LRN, Accuracy)
...
+    double l1 = 0.0, lInf = 0.0;
+    // The OpenCL kernels use the native_ math functions which have
+    // implementation defined accuracy, so we use relaxed thresholds. See
+    // https://github.com/opencv/opencv/issues/9821 for more details.
+    if (targetId == DNN_TARGET_OPENCL)
+    {
+        l1 = 0.01;
+        lInf = 0.01;
+    }
+    test(lp, input, backendId, targetId, false, l1, lInf);
```
Comment (aarongreig) [inline modules/dnn/test/test_halide_layers.cpp:264]: looks like this is going to be an issue for 3.4 as the parameterized accuracy requirements weren't implemented, should I backport that as well?

Comment (alalek) [reply modules/dnn/test/test_halide_layers.cpp:264]: yes, please backport - copy test() implementation from the master branch. It should looks like this:

```.diff
 using namespace cv::dnn;
 using namespace testing;
 
-static void test(Mat& input, Net& net, Backend backendId, Target targetId, bool skipCheck = false, bool randInput = true)
+static void test(Mat& input, Net& net, Backend backendId, Target targetId, bool skipCheck = false, bool randInput = true, double l1 = 0.0, double lInf = 0.0)
 {
     DNNTestLayer::checkBackend(backendId, targetId);
     if (randInput)
@@ -33,8 +33,12 @@ static void test(Mat& input, Net& net, Backend backendId, Target targetId, bool
     if (skipCheck)
         return;
 
-    double l1, lInf;
-    DNNTestLayer::getDefaultThresholds(backendId, targetId, &l1, &lInf);
+    double default_l1, default_lInf;
+    DNNTestLayer::getDefaultThresholds(backendId, targetId, &default_l1, &default_lInf);
+    if (l1 == 0.0)
+        l1 = default_l1;
+    if (lInf == 0.0)
+        lInf = default_lInf;
 #if 0
     std::cout << "l1=" << l1 << "  lInf=" << lInf << std::endl;
     std::cout << outputDefault.reshape(1, outputDefault.total()).t() << std::endl;
```

Comment (aarongreig) [reply modules/dnn/test/test_halide_layers.cpp:268]: should match master now

**Review by alalek — APPROVED** [review]
Well done! Thank you :+1:

**Pull request merged**